gitpod-io/gitpod
# Add workspace backup command to gpctl

## Description
Add 'workspaces backup <ws-id> to gpctl

## Related Issue(s)
Fixes #4962 

## How to test
`gpctl workspaces backup <ws-id>
`
should create a backup. It will report the relative location of the file e.g. 

`workspaces/blush-damselfly-qwj1me4f/full.tar
`

followed by the user id:

`gitpod-user-c4d13319-3e69-453c-80ae-46b760c196cf
`
This is the same format as the 'workspaces snapshot' command.

NOTE: 
The backup location is MinIO for preview envs: e.g
https://minio.rl-gpctl-ws-backup.staging.gitpod-dev.com/
Login/pwd are the access/secret keys [here](https://github.com/gitpod-io/gitpod/blob/c450d62d217eff55773f134f7d6bda07093d8a91/.werft/values.dev.yaml#L210)

For prod the backup is in [cloud storage](https://console.cloud.google.com/storage/browser?project=gitpod-191109) /gitpod-prod-user-<user id>/workspaces/...

Simulating failure conditions is tricky - we try to be as robust as possible for obvious reasons. For example, as the backup is run in a finalizer, deleting that finalizer from the pod is a logical approach. However, we reapply the finalizer on every pod event so it is re-added almost immediately.

One way to manually test end-to-end is as follows:
* make a change in the workspace (add/remove a file). 
* run the backup command
* download from minio the full.tar that is created
* undo the change
* stop the workspace
* delete the latest full.tar from minio as it no longer has the change
* upload the previously saved full.tar
* start the workspace again
* check that the original changes are present.

## Release Notes

```release-note
NONE
```
